### PR TITLE
Http-Factory

### PR DESCRIPTION
**Pull request by danopz**
I just tried to implement the PSR-17 interfaces for Slim-Http addressing #1. I copied some of the existing static methods, such as Request::createFromGlobals(). Maybe we could even call them inside the factories - else not we should remove them.
This is still not done as the classes needs some tests and there still is a todo in the ServerRequestFactory.
I just wanted to leave this here to get a review before I waste additional hours, and to start a discussion wheter we should implement them here or in an additional package.

- [x] Write some tests
- [x] call static methods or remove them
- [x] fix the todo in ServerRequestFactory
- [x] ...

### CODE REVIEW

**Comment by akrabat** [comment]
I'm in favour of supporting PSR-17

**Comment by danopz** [comment]
I pushed my last commits.

- updated factories
- moved Uri::createFromGlobals to UriFactory with `@internal` flag
- removed static factories from implementations, where PSR factories exists
- added http-interop/http-factory-tests for factory tests
- moved some tests to factory tests

Don't know how we should handle UploadedFile::createFromGlobals since the PSR doesn't handle this case we are using right now - maybe we should add a middleware for this?
Also don't know why the test suddenly fails on PHP7+?

**Comment by danopz** [comment]
Rebased to current master / fixed tests

**Comment by akrabat** [comment]
This is great work, but I'm not prepared to put it in until PSR-17 is ratified and published in the `psr` namespace.

Let's keep this around as a WIP until then.

**Comment by TakeMeNL** [comment]
Maybe starting a discussion but, is it a idea to separate the both? So Slim-Http, Slim-HttpFactory ?
Both can have there respected (dev) requirements and then can even be used seperately.
Let me know :)

**Comment by danopz** [comment]
After thinking about this I'm also in favor of creating a dedicated factory repository.
So they get their own dependencies and they will respect PHP versions 5/7 (PSR7 is PHP 5+, PSR17 is PHP 7+).

**Comment by danopz** [comment]
Updated to [psr/http-factory](https://github.com/php-fig/http-factory) 1.0.0.
Tests are failing because of an increased PHP version + #54

**Comment by danopz** [comment]
@akrabat should we create an extra package to keep dependencies clean?

**Comment by akrabat** [comment]
Maybe we make Slim\Http a PHP 7+ library?

**Comment by akrabat** [comment]
This needs rebasing to pick up the changes that make Slim-Http PHP 7+

**Comment by danopz** [comment]
Rebased :+1:

**Comment by johnrobertporter** [comment]
I'm not in favour of splitting this to a separate package. It's all directly related, i.e. you can't use the factories without the messages.

**Comment by johnrobertporter** [comment]
Also, can we not make a single class for `RequestFactory` and `ServerRequestFactory` implementing both interfaces? It doesn't need to be separate as it's final output is the same class; it makes sense to me to have a single factory for a single class type.

File: src/Factory/RequestFactory.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+     * @param UriInterface|string $uri The URI associated with the request. If
+     *     the value is a string, the factory MUST create a UriInterface
+     *     instance based on it.
+     *
+     * @return RequestInterface
+     */
+    public function createRequest(string $method, $uri): RequestInterface
+    {
+        if (is_string($uri)) {
+            $uri = (new UriFactory())->createUri($uri);
```
Comment (johnrobertporter) [inline src/Factory/RequestFactory.php]: I'm not comfortable with this. A Factory class is an injectable dependency, not a newable class. Should this not be injected via a constructor?

File: src/Factory/RequestFactory.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        } elseif (!$uri instanceof UriInterface) {
+            throw new \InvalidArgumentException('URI must either be string or instance of ' . UriInterface::class);
+        }
+
+        $body = (new StreamFactory())->createStream();
```
Comment (johnrobertporter) [inline src/Factory/RequestFactory.php]: Again as with `UriFactory`.

File: src/Factory/ServerRequestFactory.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+     *     instance based on it.
+     * @param array $serverParams Array of SAPI parameters with which to seed
+     *     the generated request instance.
+     *
+     * @return ServerRequestInterface
+     */
+    public function createServerRequest(string $method, $uri, array $serverParams = []): ServerRequestInterface
+    {
+        if (is_string($uri)) {
+            $uri = (new UriFactory())->createUri($uri);
```
Comment (johnrobertporter) [inline src/Factory/ServerRequestFactory.php]: Again, an injectable dependency IMO.

File: src/Factory/ServerRequestFactory.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        } elseif (!$uri instanceof UriInterface) {
+            throw new \InvalidArgumentException('URI must either be string or instance of ' . UriInterface::class);
+        }
+
+        $body = (new StreamFactory())->createStream();
```
Comment (johnrobertporter) [inline src/Factory/ServerRequestFactory.php]: Again, an injectable dependency IMO.

File: src/Factory/UriFactory.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+     */
+    public function createUri(string $uri = ''): UriInterface
+    {
+        $parts = parse_url($uri);
+
+        if ($parts === false) {
+            throw new \InvalidArgumentException('URI cannot be parsed');
+        }
+
+        $scheme = isset($parts['scheme']) ? $parts['scheme'] : '';
```
Comment (johnrobertporter) [inline src/Factory/UriFactory.php]: If we're using PHP 7, could we do something like:

```php
$scheme = $parts['scheme'] ?? '';
```

**Comment by akrabat** [comment]
> Also, can we not make a single class for `RequestFactory` and `ServerRequestFactory` implementing both interfaces? It doesn't need to be separate as it's final output is the same class; it makes sense to me to have a single factory for a single class type.

Should we have a `Request` and `ServerRequest` in Slim-Http?

File: src/Factory/RequestFactory.php (hunk 1 above)
Comment (danopz) [reply src/Factory/RequestFactory.php]: Yeah I agree, so one could inject some other factory and must not live with ours.

**Comment by danopz** [comment]
Now we can inject a StreamFactoryInterface and UriFactoryInterface in both RequestFactories defaulting to null and so to our factories.
I'm not sure about merging RequestFactory and ServerRequestFactory.

**Comment by akrabat** [comment]
This needs to be against [Slim-Psr7](https://github.com/slimphp/Slim-Psr7) now. 

https://github.com/slimphp/Slim-Psr7/issues/14 tracks it.

**Pull request closed without merging**